tools-local/unix/tool-helper.sh

```diff
@@ -0,0 +1,272 @@
+
+    toolFolder="$(get_tool_config_value "$repoRoot" "$toolName" "LocalToolFolder")"
+
+    if [ $? -ne 0 ]; then
+        toolFolder="Tools/downloads/$toolName"
+    fi
+
+    case "$toolFolder" in
+        /*)
+            echo "$toolFolder"
+            ;;
+        *)
+            # Assumed that the path specified in .toolversion is relative to the repository root.
+            echo "$repoRoot/$toolFolder"
+            ;;
+    esac
+}
+
+# Gets the search path corresponding to the specified tool name.
+# Search path is read from the .toolversions file.
+# Exit 1 if unable to read the path from the .toolversions file.
+get_local_search_path()
+{
+    repoRoot="$1"
+    toolName="$2"
+
+    exit_if_arg_empty "repository-root" "$repoRoot"
+    exit_if_arg_empty "tool-name" "$toolName"
+
+    toolFolder="$(get_local_tool_folder "$repoRoot" "$toolName")"
+    osName="$(get_os_name)"
+    searchPath="$(get_tool_config_value "$repoRoot" "$toolName" "LocalSearchPath${osName}")"
+    echo "$toolFolder/$searchPath"
+}
+
+# Gets the error message to be displayed when the specified tool is not available for the build.
+# Error message is read from the .toolversions file.
+# Exit 1 if unable to read the error message from the .toolversions file.
+tool_not_found_message()
+{
+    repoRoot="$1"
+    toolName="$2"
+
+    exit_if_arg_empty "repository-root" "$repoRoot"
+    exit_if_arg_empty "tool-name" "$toolName"
+
+    # Eval in a subshell to avoid conflict with existing variables.
+    (
+        eval_tool "$repoRoot" "$toolName"
+
+        if [ -z "$ToolNotFoundMessage" ]; then
+            echo "Unable to locate $toolName."
+            exit 1
+        fi
+
+        eval echo "$ToolNotFoundMessage"
+    )
+}
+
+# Write the given message(s) to probe log file.
+log_message()
+{
+    repoRoot="$1"
+    exit_if_arg_empty "repository-root" "$repoRoot"
+    probeLog="$repoRoot/probe-tool.log"
+    shift
+    echo "$*" >> "$probeLog"
+}
+
+# Displays and logs the specified message, and exits with status 1.
+fail()
+{
+    repoRoot="$1"
+    exit_if_arg_empty "repository-root" "$repoRoot"
+    log_message "$@"
+    shift
+    echo "$@"
+    exit 1
+}
+
+# Invokes the override extension script if available, else invokes the base implementation.
+invoke_extension()
+{
+    # Displays the usage for invoke_extension function.
+    invoke_extension_usage()
+    {
+        echo "usage: invoke_extension <script-name> <repository-root> <tool-name> <override-scripts-folder-path> [args ...]"
+        echo "script-name                       Name of the extension script."
+        echo "repository-root                   Path to repository root."
+        echo "tool-name                         Name of the tool."
+        echo "override-scripts-folder-path      If a path is specified then, search and acquire scripts from the specified folder will be invoked. Otherwise, search will use the default search and acquire scripts located within the repository."
```

Comment:
This line (when echoed) is 230 chars long! This would be hard to read on an 80-wide shell (typically considered standard). `dotnet-install.sh` is guilty of 116 in its usage doc, so I don't know if anyone cares about this nowadays.